pkgs/native_assets_cli/lib/src/api/build_state.dart

```diff
@@ -82,27 +84,51 @@ final class BuildState {
     await output.writeToFile(outDir: config.outDir);
   }
 
-  final BuildConfig config;
-  final BuildOutput output;
+  final BuildConfig _config;
+  final BuildOutput _output;
 
   BuildState._({
-    required this.config,
-    required this.output,
-  });
-
-  /// Adds assets to build output.
-  ///
-  /// See [BuildOutput.addAssets] for more info.
-  void addAssets(Iterable<Asset> assets) => output.addAssets(assets);
+    required BuildConfig config,
+    required BuildOutput output,
+  })  : _output = output,
+        _config = config;
 
   /// Adds dependencies to build output.
   ///
   /// See [BuildOutput.addDependencies] for more info.
   void addDependencies(Iterable<Uri> dependencies) =>
-      output.addDependencies(dependencies);
+      _output.addDependencies(dependencies);
 
   /// Adds metadata to build output.
   ///
   /// See [BuildOutput.addMetadata] for more info.
-  void addMetadata(String key, Object value) => output.addMetadata(key, value);
+  void addMetadata(String key, Object value) => _output.addMetadata(key, value);
+
+  /// Adds asset to build output.
+  ///
+  /// See [BuildOutput.addAssets] for more info.
+  void addAsset({required String id, required AssetPath path}) {
```

Comment:
Instead of adding methods to BuildState or BuildOutput, I was thinking to add a `AssetDownloader` similar to `CBuilder`.

Let me try to add that to the redesign PR.